consumergroup/offset_manager.go

```diff
@@ -0,0 +1,155 @@
+package consumergroup
+
+import (
+	"sync"
+	"time"
+)
+
+type OffsetManager interface {
+	InitializePartition(topic string, partition int32) (int64, error)
+	MarkAsSeen(topic string, partition int32, offset int64) bool
+	Close() error
+}
+
+type OffsetManagerConfig struct {
+	CommitInterval time.Duration
+	VerboseLogging bool
+}
+
+type ZookeeperOffsetManager struct {
+	config  *OffsetManagerConfig
+	l       sync.RWMutex
+	offsets offsetsMap
+	cg      *ConsumerGroup
+
+	closing chan struct{}
+	closed  chan struct{}
+}
+
+func NewZookeeperOffsetManager(cg *ConsumerGroup, config *OffsetManagerConfig) OffsetManager {
+	if config == nil {
+		config = &OffsetManagerConfig{CommitInterval: 10 * time.Second}
+	}
+
+	zom := &ZookeeperOffsetManager{
+		config:  config,
+		cg:      cg,
+		offsets: make(offsetsMap),
+		closing: make(chan struct{}),
+		closed:  make(chan struct{}),
+	}
+
+	go zom.offsetCommitter()
+
+	return zom
+}
+
+func (zom *ZookeeperOffsetManager) MarkAsSeen(topic string, partition int32, offset int64) bool {
+	zom.l.RLock()
+	defer zom.l.RUnlock()
+	return zom.offsets[topic][partition].MarkAsSeen(offset)
+}
+
+func (zom *ZookeeperOffsetManager) InitializePartition(topic string, partition int32) (int64, error) {
+	zom.l.Lock()
+	defer zom.l.Unlock()
+
+	if zom.offsets[topic] == nil {
+		zom.offsets[topic] = make(topicOffsets)
+	}
+
+	nextOffset, err := zom.cg.zk.Offset(zom.cg.name, topic, partition)
+	if err != nil {
+		return -1, err
+	}
+	zom.offsets[topic][partition] = newPartitionOffsetTracker(nextOffset - 1)
+
+	return nextOffset, nil
+}
+
+func (zom *ZookeeperOffsetManager) Close() error {
+	close(zom.closing)
+	<-zom.closed
+	return zom.commitOffsets()
+}
+
+func (zom *ZookeeperOffsetManager) offsetCommitter() {
+	commitTicker := time.NewTicker(zom.config.CommitInterval)
+	defer commitTicker.Stop()
+
+	for {
+		select {
+		case <-zom.closing:
+			close(zom.closed)
+			return
+		case <-commitTicker.C:
+			zom.commitOffsets()
+		}
+	}
+}
+
+func (zom *ZookeeperOffsetManager) commitOffsets() error {
+	zom.l.RLock()
+	defer zom.l.RUnlock()
+
+	var returnErr error
+	for topic, partitionOffsets := range zom.offsets {
+		for partition, offsetTracker := range partitionOffsets {
+			committer := func(offset int64) error {
+				return zom.cg.zk.Commit(zom.cg.name, topic, partition, offset+1)
+			}
+
+			err := offsetTracker.Commit(committer)
+			switch err {
+			case nil:
+				if zom.config.VerboseLogging {
+					zom.cg.Logf("Committed offset %d for %s:%d\n", offsetTracker.highestOffsetSeen, topic, partition)
+				}
+			case AlreadyCommitted:
+				// noop
+			default:
+				returnErr = err
+				zom.cg.Logf("Failed to commit offset %d for %s:%d\n", offsetTracker.highestOffsetSeen, topic, partition)
+			}
+		}
+	}
+	return returnErr
+}
+
+type offsetCommitter func(int64) error
+type topicOffsets map[int32]*partitionOffsetTracker
+type offsetsMap map[string]topicOffsets
```

Comment:
the C programmer in me wants these at the top of the file, above all their usages :)